indexbs/indexbacked_bs.go

```diff
@@ -0,0 +1,194 @@
+type ShardSelectorF func(c cid.Cid, shards []shard.Key) (shard.Key, error)
+
+type accessorWithBlockstore struct {
+	sa *dagstore.ShardAccessor
+	bs dagstore.ReadBlockstore
+}
+
+// IndexBackedBlockstore is a read only blockstore over all cids across all shards in the dagstore.
+type IndexBackedBlockstore struct {
+	d            *dagstore.DAGStore
+	shardSelectF ShardSelectorF
+
+	// caches the blockstore for a given shard for shard read affinity i.e. further reads will likely be from the same shard.
+	// shard key -> read only blockstore
+	blockstoreCache *lru.Cache
+}
+
+func NewIndexBackedBlockstore(d *dagstore.DAGStore, shardSelector ShardSelectorF, maxCacheSize int, maxBlocks int) (blockstore.Blockstore, error) {
+	// instantiate the blockstore cache
+	bslru, err := lru.NewWithEvict(maxCacheSize, func(_ interface{}, val interface{}) {
+		// ensure we close the blockstore for a shard when it's evicted from the cache so dagstore can gc it.
+		abs := val.(*accessorWithBlockstore)
+		abs.sa.Close()
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create lru cache for read only blockstores")
+	}
+
+	return &IndexBackedBlockstore{
+		d:               d,
+		shardSelectF:    shardSelector,
+		blockstoreCache: bslru,
+	}, nil
+}
+
+func (ro *IndexBackedBlockstore) Get(ctx context.Context, c cid.Cid) (b blocks.Block, finalErr error) {
+	logbs.Debugw("Get called", "cid", c)
+	defer func() {
+		if finalErr != nil {
+			logbs.Debugw("Get: got error", "cid", c, "error", finalErr)
+		}
+	}()
+
+	mhash := c.Hash()
+
+	// fetch all the shards containing the multihash
+	shards, err := ro.d.ShardsContainingMultihash(ctx, mhash)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch shards containing the block: %w", err)
+	}
+	if len(shards) == 0 {
+		return nil, ErrBlockNotFound
+	}
+
+	// do we have a cached blockstore for a shard containing the required block ? If yes, serve the block from that blockstore
+	for _, sk := range shards {
+		// a valid cache hit here updates the priority of the shard's blockstore in the LRU cache.
+		val, ok := ro.blockstoreCache.Get(sk)
+		if !ok {
+			continue
+		}
+
+		rbs := val.(*accessorWithBlockstore).bs
+		blk, err := rbs.Get(ctx, c)
+		if err != nil {
+			// we know that the cid we want to lookup belongs to a shard with key `sk` and
+			// so if we fail to get the corresponding block from the blockstore for that shards, something has gone wrong
+			// and we should remove the blockstore for that shard from our cache.
+			ro.blockstoreCache.Remove(sk)
+			continue
+		}
+
+		// add the block to the block cache
+		logbs.Debugw("Get: returning from block store cache", "cid", c)
+		return blk, nil
+	}
+
+	// ---- we don't have a cached blockstore for a shard that can serve the block -> let's build one.
+
+	// select a valid shard that can serve the retrieval
+	sk, err := ro.shardSelectF(c, shards)
+	if err != nil {
+		return nil, ErrBlockNotFound
```

Comment:
I think we may want to distinguish between a "not found" error, and other errors returned from shardSelectF

Reply:
Done.